### PR TITLE
kafka: new deserializers

### PR DESCRIPTION
**Pull request by adamkotwasinski**
Commit Message: kafka: new deserializers
Additional Description: new deserializers for kafka variable-length encoded int32 & int64 (called [varint, varlong](https://kafka.apache.org/protocol.html#protocol_types) there). These deserializers will be used in kafka-mesh-filter to decode internals of received kafka produce request objects so we can get the real messages out of the envelope [(like here)](https://github.com/envoyproxy/envoy/pull/11936/files#diff-a5b85c101e48a968f816536f17e9e7ab84db7abf19e468826f30085260ad2421R199), and forward them to kafka producers.
Risk Level: Low
Testing: unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/extensions/filters/network/kafka/serialization.h [hunk 1]
```diff
@@ -224,6 +224,72 @@ class VarUInt32Deserializer : public Deserializer<uint32_t> {
   bool ready_ = false;
 };
 
+class VarInt32Deserializer : public Deserializer<int32_t> {
```
Comment (junr03) [inline source/extensions/filters/network/kafka/serialization.h:227]: should we add class comments like the other classes?

File: source/extensions/filters/network/kafka/serialization.h [hunk 2]
```diff
@@ -224,6 +224,72 @@ class VarUInt32Deserializer : public Deserializer<uint32_t> {
...
+  int32_t get() const override {
+    const uint32_t res = varuint32_deserializer_.get();
+    return (res >> 1) ^ -(res & 1);
+  }
+
+private:
+  VarUInt32Deserializer varuint32_deserializer_;
+};
+
+class VarInt64Deserializer : public Deserializer<int64_t> {
```
Comment (junr03) [inline source/extensions/filters/network/kafka/serialization.h:244]: Do we eventually need a VarUInt64Deserializer that this deserializer can be written on top off like the above?

File: test/extensions/filters/network/kafka/serialization_test.cc [hunk 3]
```diff
@@ -151,7 +154,76 @@ TEST(VarUInt32Deserializer, ShouldThrowIfNoEndWith5Bytes) {
 
   // when
   // then
-  EXPECT_THROW(testee.feed(data), EnvoyException);
+  EXPECT_THROW_WITH_REGEX(testee.feed(data), EnvoyException, "is too long");
+}
+
+// Missing tests (chunks).
```
Comment (junr03) [inline test/extensions/filters/network/kafka/serialization_test.cc:160]: is this a todo?

File: source/extensions/filters/network/kafka/serialization.h (hunk 2 above)
Comment (adamkotwasinski) [reply source/extensions/filters/network/kafka/serialization.h:244]: No, fortunately not.

**Comment by adamkotwasinski** [comment]
/wait let's wait for the builds first

File: test/extensions/filters/network/kafka/serialization_test.cc (hunk 3 above)
Comment (adamkotwasinski) [reply test/extensions/filters/network/kafka/serialization_test.cc]: Yes - sorry about that! More tests have been added.

File: source/extensions/filters/network/kafka/serialization.h (hunk 1 above)
Comment (adamkotwasinski) [reply source/extensions/filters/network/kafka/serialization.h:237]: Great idea!

**Review by junr03 — COMMENTED** [review]
thanks for the changes will move up to @mattklein123 for final pass.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**